src/extjson/de.rs

```diff
@@ -0,0 +1,230 @@
+//! Deserializing [MongoDB Extended JSON v2](https://docs.mongodb.com/manual/reference/mongodb-extended-json/)
+//!
+//! ## Usage
+//!
+//! Extended JSON can be deserialized using [`Bson`](../../enum.Bson.html)'s
+//! `TryFrom<serde_json::Value>` implementation. This implementation accepts both canonical and
+//! relaxed extJSON, and the two modes can even be mixed within a single representation.
+//!
+//! e.g.
+//! ```rust
+//! # use bson::Bson;
+//! # use serde_json::json;
+//! # use std::convert::{TryFrom, TryInto};
+//! let json_doc = json!({ "x": 5i32, "y": { "$numberInt": "5" }, "z": { "subdoc": "hello" } });
+//! let bson: Bson = json_doc.try_into().unwrap(); // Bson::Document(...)
+//!
+//! let json_date = json!({ "$date": { "$numberLong": "1590972160292" } });
+//! let bson_date: Bson = json_date.try_into().unwrap(); // Bson::DateTime(...)
+//!
+//! let invalid_ext_json = json!({ "$numberLong": 5 });
+//! Bson::try_from(invalid_ext_json).expect_err("5 should be a string");
+//! ```
+
+use std::convert::{TryFrom, TryInto};
+
+use serde::de::{Error as _, Unexpected};
+
+use crate::{extjson::models, oid, Bson, Document};
+
+#[derive(Debug)]
+#[non_exhaustive]
+/// Error cases that can occur during deserialization from [extended JSON](https://docs.mongodb.com/manual/reference/mongodb-extended-json/).
+pub enum Error {
+    /// Errors that can occur during OID construction and generation from the input data.
+    InvalidObjectId(oid::Error),
+
+    /// A general error encountered during deserialization.
+    /// See: https://docs.serde.rs/serde/de/trait.Error.html
+    DeserializationError { message: String },
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            Self::InvalidObjectId(ref err) => err.fmt(fmt),
+            Self::DeserializationError { ref message } => message.fmt(fmt),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl serde::de::Error for Error {
+    fn custom<T>(msg: T) -> Self
+    where
+        T: std::fmt::Display,
+    {
+        Self::DeserializationError {
+            message: format!("{}", msg),
+        }
+    }
+}
+
+impl From<serde_json::Error> for Error {
+    fn from(err: serde_json::Error) -> Self {
+        Self::DeserializationError {
+            message: err.to_string(),
+        }
+    }
+}
+
+impl From<oid::Error> for Error {
+    fn from(err: oid::Error) -> Self {
+        Self::InvalidObjectId(err)
+    }
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// This converts from the input JSON object as if it were [MongoDB Extended JSON v2](https://docs.mongodb.com/manual/reference/mongodb-extended-json/).
+impl TryFrom<serde_json::Map<String, serde_json::Value>> for Bson {
```

Comment:
this method could be simplified greatly through clever uses of macros and traits or something, but for the time being I just wanted to keep it simple, with the idea being we can restructure it after we ship 1.0